go/externals/proof_service_generic_social.go

```diff
@@ -45,48 +23,178 @@ type GenericSocialProofConfig struct {
+	url = strings.Replace(url, sigHashKey, sigID.String(), 1)
 	if !strings.Contains(url, sigID.String()) {
-		return "", fmt.Errorf("Invalid PrefillUrl, missing sigHash")
+		return "", fmt.Errorf("Invalid PrefillUrl %s, missing sigHash", url)
+	}
+	return url, nil
+}
+
+func (c *GenericSocialProofConfig) checkURLWithValues(remoteUsername string) (string, error) {
+	url := strings.Replace(c.CheckUrl, remoteUsernameKey, remoteUsername, 1)
+	if !strings.Contains(url, remoteUsername) {
+		return "", fmt.Errorf("Invalid CheckUrl %s, missing username", url)
 	}
 	return url, nil
 }
 
+//=============================================================================
+// GenericSocialProof
+//
+
+type GenericSocialProofChecker struct {
+	proof  libkb.RemoteProofChainLink
+	config *GenericSocialProofConfig
+}
+
+var _ libkb.ProofChecker = (*GenericSocialProofChecker)(nil)
+
+func NewGenericSocialProofChecker(proof libkb.RemoteProofChainLink, config *GenericSocialProofConfig) (*GenericSocialProofChecker, libkb.ProofError) {
+	return &GenericSocialProofChecker{
+		proof:  proof,
+		config: config,
+	}, nil
+}
+
+func (rc *GenericSocialProofChecker) GetTorError() libkb.ProofError { return nil }
+
+func (rc *GenericSocialProofChecker) CheckStatus(mctx libkb.MetaContext, _ libkb.SigHint, _ libkb.ProofCheckerMode,
+	pvlU keybase1.MerkleStoreEntry) (*libkb.SigHint, libkb.ProofError) {
+
+	_, sigID, err := libkb.OpenSig(rc.proof.GetArmoredSig())
+	if err != nil {
+		return nil, libkb.NewProofError(keybase1.ProofStatus_BAD_SIGNATURE,
+			"Bad signature: %v", err)
+	}
+
+	remoteUsername := rc.proof.GetRemoteUsername()
+	apiURL, err := rc.config.checkURLWithValues(remoteUsername)
+	if err != nil {
+		return nil, libkb.NewProofError(keybase1.ProofStatus_BAD_API_URL,
+			"Bad api url: %v", err)
+	}
+
+	if _, err = url.Parse(apiURL); err != nil {
+		return nil, libkb.NewProofError(keybase1.ProofStatus_FAILED_PARSE,
+			"Could not parse url: '%v'", apiURL)
+	}
+
+	res, err := mctx.G().GetExternalAPI().Get(libkb.APIArg{
+		Endpoint:    apiURL,
+		MetaContext: mctx,
+	})
+	if err != nil {
+		return nil, libkb.XapiError(err, apiURL)
+	}
+
+	// We expect a single result to match which contains an array of proofs.
+	results, perr := libkb.AtSelectorPath(res.Body, rc.config.CheckPath, mctx.CDebugf)
+	if perr != nil {
+		return nil, perr
+	}
+	if len(results) != 1 {
+		return nil, libkb.NewProofError(keybase1.ProofStatus_CONTENT_FAILURE,
+			"Json selector did not match any values")
+	}
+	var proofs []keybase1.ParamProofJSON
+	if err = results[0].UnmarshalAgain(&proofs); err != nil {
+		return nil, libkb.NewProofError(keybase1.ProofStatus_CONTENT_FAILURE,
+			"Json could not be deserialized")
+	}
+
+	found := false
+	for _, proof := range proofs {
+		if proof.KbUsername == rc.proof.GetUsername() && sigID.Equal(proof.SigHash) {
+			found = true
+			break
+		}
+	}
+	if !found {
+		return nil, libkb.NewProofError(keybase1.ProofStatus_NOT_FOUND,
+			"Unable to find the proof")
+	}
```

Comment:
it might be nice for debugging and maybe for users to have different error messages for when username isn't found vs when the username is found but the sig id is wrong.